### ISSUE TITLE
Some Playwright types now reported as `any` after upgrading to 5.7.2

### ISSUE THREAD

**Issue by mrmckeb** [microsoft/TypeScript#60619]
### 🔎 Search Terms

I manually scrolled through all issues reported since the release of 5.7.2.

### 🕗 Version & Regression Information

- This changed between versions 5.6.3 and 5.7.2

### ⏯ Playground Link

https://github.com/mrmckeb/pw-repro

### 💻 Code

Minimal repro available here:
https://github.com/mrmckeb/pw-repro

The Playwright team are also able to reproduce:
https://github.com/microsoft/playwright/issues/33763

The code is this:
```ts
import { test as base, expect } from "@playwright/test";

const test = base.extend({
  // baseUrl, page, use are now `any`
  page: async ({ baseURL, page }, use) => {
    await page.goto(baseURL);
    await use(page);
  },
});
```

### 🙁 Actual behavior

Types that were previously available are now typed as `any`. JSDoc continues to work as expected.

### 🙂 Expected behavior

Types behave as they did in TypeScript 5.6.3.

### Additional information about the issue

We found this issue when upgrading from 5.6.3 to 5.7.2. I couldn't find a related issue here, so raised it at the Playwright repo to at least verify that they could reproduce and that they didn't think it was an issue with their types.

The Playwright team also confirmed that they have 71 new errors across 41 files when moving to TypeScript 5.7.2.

**Comment by Andarist** [comment]
I don't understand why `Exclude<R, Function>` is there but I can also see how it could potentially mess things up (although I wasn't yet able internally verify why it messes it up). Unless I'm missing some constraint that those types try to apply there I think this is actually incorrect.

I'll dive deeper into it later but in the meantime I've also created a different repro case that doesn't include that `Exclude` and I think it's better to focus on it first: [TS playground](https://www.typescriptlang.org/play/?ts=5.8.0-dev.20241127#code/C4TwDgpgBA0hIDUCGAbArtAvFA3lA2gNbwBcUAzsAE4CWAdgOYC6ZSdIUAvgNwBQvEAB5gA9lWBRQkKABUIlAGI1BwNFQgAeAEoAaKAEEqDclCHAIdACYm4iVBgB8UbAApeUKEiPkyh4zvcoNHIIMhcqMi0ASmcnAAUqEQBbGhCNADcRGksHAJjMJzYQPl4paCUVNXkNGVMVC2tYeGR0LFxOPTjaswabZvs2nE4nbBxA-BgoeihiEBEAMygupgB+EkCPAB9ZeWAK1XUNLommPVqAMiWZBw2obfw5RWUDzWOYU9koS67c3EkaJIQERoYBrKB0NBJABGECodyCVgg83oEEsXCYfE4X1w40m01mC1kq3WHi2OyelUOMhOZ2xP1u90ee2eVRqNM+32uejwwABQJBYIh0Nh8LQiORdFR6MxJXo5io8yQAGNoEyZOBNEy-CYelY+nZWk4xh5dZYanVzHqmgbHG5SVBkZT5GR9lVyDUzrttblAlEyGqNTUvd5sddMfw5bDFSqlkgGNBjaZ0vYkOYwpQIhRqPQGPl4okUmkxYQ6CIAO50Bzh3iWCBKlBeaBKkR0ShQKFIEL+3bqyAaPBgOOhWPxrhV-gdkIAOlNLkTg-jrHIIDoSqgc6gC+gnDz7QCO5KQlE4nafCAA)

**Comment by dgozman** [comment]
@Andarist The `Exclude<R, Function>` construct is to allow the fixture value in a literal form, e.g. `page: 'foo'`, in addition to the fixture function like `page: async ({}, use) => use('foo')`. However, if the fixture type `F` is a function, the idea is to only allow the latter form, to avoid any possibility of the confusion between `F` and `async ({}, use) => {...}`.

By the way, I once tried to replace `[K in keyof T]?:` with `[K in Exclude<keyof T, keyof PT>]?:` to avoid the clash between `T` and `PT` properties. However, that breaks when `R` or `Exclude<R, Function>` is allowed in the `TestFixture` type, because it infers `T` in `extend<T extends KeyValue>` as `KeyValue` and every property becomes `any`.

**Comment by Andarist** [comment]
> @Andarist The Exclude<R, Function> construct is to allow the fixture value in a literal form, e.g. page: 'foo', in addition to the fixture function like page: async ({}, use) => use('foo'). However, if the fixture type F is a function, the idea is to only allow the latter form, to avoid any possibility of the confusion between F and async ({}, use) => {...}.

Ah ye, I already figured this part out (side note - you don't have any type tests in Playwright for this scenario, no test fail after replacing `Exclude<R, Function>` with just `R` there).

I'm not yet sure where it breaks internally but, in general, "not a type" isn't something greatly supported by TS. I put together some notes~ on various ways to do it in this [TS playground](https://www.typescriptlang.org/play/?ts=5.8.0-dev.20241127#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwxAGcMBGAHgBUA+ACgDcAueS+AH3ltuSJGYeaUAlPAC81ePRxZgI8ZOmyhggNwAoNWDwl4cIsghkxBYmS6wA5s1TIAtgCMQMORIDeAXyHqtqHXoMYAEzGhCQUtJbWdo7OYm7udBEwFi7wHl7wAPSZ8LZQANbE8Ly+IAA08ADuCAAWUPQgqADkGPDAWERQDlgWyFCEGqCQsAgo6Nh4JiSBVHRq8JKM8wvs8ACiAB6QyKBUFQBiaJi4qNTLCxxcPHycWKiIOMyu8AAOUBY3aPmoOJX47hUmCxUlIZCDFEIymplCx1JptK1-IYAMwhUyBczJKIOJypdIqLI5KD4EAbF7gQjAeBOGA4GDwO4EGodYpYDB9CaoDTZJkslmoLBgEAQACe8AARO0iGAYFhbHd+iBgOKCDgcPDfIjiAEACxo6aYlJxNLLSzLHkAPQA-GpPOoeTh8hUQA0YCKMMzUBYGUQkFgNkqAHTB+D2ZCtHh3b1QeA+dqcqAQAgi8nwYlUyo1RqvWn0GRR2N4QgbdmJ16wLogQj0jApops3I9GqtEhYCBJxxtECIGlKjV+bWGACs+qC5CSVngNhxsXkHkSljxdsG4GgcELmqnOAwAEFDuMTsw0FgAI7IBBEEUOHAQdR3auIKBC+D746TVzLADaP13r85AF0rSPVBvl+VB1HcDRa1TAA5bc9yOTljA-BZv3gv8TkA6wXScCC4R5D0+V9GN2kQHs4FQFsMBgRULDFB56XFH80yQRCTmTclxTUIY11GNjJlCDBkSoakS0aYBfTg39+NOWhliBVhLm4Xh+DuB4nnLD5gNAv54ABRZgWNUFZCMiEoRhSg4R8Ad9EMAA2UdkUNbEYiXDIeWJUTyUwJVqRgWl6UZQjfSINkORObkcmCn0p0FYUxUlDoZTlBVKRVDA1X7LVbIwAB2RzDTxU1knNHJrVtLw1AI5lfX5bcmQQUBQrgKkcHDLRbFqJxyjTX0xjfXxihqNqICpVt21DBBEwgX4g2DKqcgbSoRqpAB5ABpVU2iS2AqX6zliMMHAPg9Jx4AASX2BqxWWgwqU7CwsAafAMt5PqZOKHAGVaW7RtjabvoWlj9rwJpatsF4IBATrKP6diwGJFpts6bpekVSaMGqbN9vY6Cin6XQ0GwTq00Qas03bIGw1aIKatZUskNO1iD0mFlMwJpnrnpBH8DeIhfVegADa5BbTVAqWihH23sJ98hixNKigEVfQgLB7Bot0AFoXlzGQ+2s7KAgADkc8dIinaJcWNedCut5cgeilkRSrMXtyzGtKg6VoBKzNN7BwBpuNXEZmYGqYMB1WY5IWJhzlWWhWAAMngKSEJZ1AhDjpTrlU+5HjSTTPhAn5dP0hTwTBUywXM1QNAN3RBwwABOUcdWci2ZzcgkPJJMkKV8mk6QZF66dChmIodun+Ti0UJSlZL5VQRVlVVdV66RMgAAZW9tudiosUr4HK5ceWp+A0AY9kl8IUVg0Db62hwYhmlaRB-TerKG5y0hSFbs2sQ7q5O2C5khdzUEAA). Ultimately, only your strategy gives the desired results here.

> By the way, I once tried to replace [K in keyof T]?: with [K in Exclude<keyof T, keyof PT>]?: to avoid the clash between T and PT properties. However, that breaks when R or Exclude<R, Function> is allowed in the TestFixture type, because it infers T in extend<T extends KeyValue> as KeyValue and every property becomes any.

You are on the right track here when it comes to the root cause. The problem is that your `PT` is already a resolved mapped type with function properties. But then `T` is being inferred into *from the same object that is typed using an intersection over mapped types applied to T and PT* and it really doesn't know that it shouldn't use `page` property because it's already part of `PT`.

With my change the compiler uses all property types that could come out of deferred instantiations at this position. So it uses function types from all of those 4 mapped types and they are just not compatible with each other. That's how we end up with implicit anys.

To be fair, even before TS 5.7 the inference didn't quite work OK in your case (the contextual parameter types did though): [TS playground](https://www.typescriptlang.org/play/?ts=5.6.3#code/JYWwDg9gTgLgBAbzjApgZ3gQzXARtlAGjhQA8wUBjeAXzgDMoIQ4AiAATABtMBPAdyjAA5gAsYAelQZWAbjgSJyXhTQAuOAEYAdABYAnNoAMAKBOK4wAHb0UUHAANtZVFYAmAHgDSKXgDVMLgBXIkQaAD4HE3w0FGdSVzcACgQTOAUlAD0AfjS4MExhFA1sXitKOCSAfQBKOABecLDCExoa2TMLa1t7OCcXFHdvXwDg0IQIqJi4gfcUvIscvJBeADFgBKCoYrhS8srahqaJlraO8yVuu0d4xOH-QJDiCcjoglvB5NT0xdz0gqKJTQZQq1TqjWayzWGxgWx2e1BhwhJ1a7U6lxs1wufQ+Q1SFnS-0KOyCVgA1lYIPwrB0CYSVutNtsNElwU0AApMEDAWIeABuEGAbnCtKUNGe2MJ+WJGlJFKpNMl9OhTJJ5Mp1NFcEmb1iuK+Cyyf2lgN2wP2YKOzQycEoECsqASAFpYlY0MAYMA+SgoYzYcyzSDKmzrRY7Q6XE7rK73Z7vajzl1MfZsf0Ep8PPilFKATtWVbOcweSh+YLhVqpQyYXCWSHC9zeQKhSLseLEEqiaa5RrFXT0lXVbL1QqtTrpvr5j8jXlc0Cg-nkW2w-bHTAo27BrGvT7+yr-fDzaCQycbeHV+uYx7twn0ZZk2hU-rMx24AP97WC1zi6Xm6OJX3Xz3Gs4G7EdW1ecdZgNKc4CWXc-WAhFgytE9lwjZ1o03K94zOIA). I somewhat doubt you care about it, when defining extra fixtures it's advised to use type params explicitly here instead of relying on the inference. It would certainly be great to make it work though. 

Luckily-ish since TS 5.6 (since [this PR](https://github.com/microsoft/TypeScript/pull/56201)) we can leverage filtering mapped types to disambiguite slightly between those mapped types: [TS playground](https://www.typescriptlang.org/play/?ts=5.7.2#code/C4TwDgpgBA0hIDUCGAbArtAvFA3lA2gNbwBcUAzsAE4CWAdgOYC6ZSdIUAvgNwBQvEAB5gA9lWBRQkKABUIlAGI1BwNFQgAeAEoAaKAEEqDclCHAIdACYm4iVBgB8UbAApeUKEiPkyh4zvcoNHIIMhcqMi0ASmcnAAUqEQBbGhCNADcRGksHAJjMJzYQPl4paCUVNXkNGVMVC2tYeGR0LFxOPTjaswabZvs2nE4nbBxA-BgoeihiEBEAMygupgB+EkCPAB9ZeWAK1XUNLommPVqAMiWZBw2obfw5RWUDzWOYU9koS67c3EkaJIQERoYBrKB0NBJABGECodyCVgg83oEEsXCYfE4X1wgQA9LiKBBoMAABapTwAd3kyWgyMEAEIoCQouNJtNZgtPkgTABRQQAY3Qlk0MD0HMWP1W6w8Wx2T0qhxkJzO2J+t3ujz2zyqNWVn2+1z0eGAAKBILBEOhsPhaERyLoqPRmJK9HMVHmSH50E1MnAmk1fhMPSsfTsrScYw8wcsNTq5hDTTDjjcMqgdJePig+yq5BqZ12gdygSiZB9fpqBe82OumP4rthHq9SyQDGgkdM6XsSHMYUoEQo1HoDHy8USKTStsIdBEFLoDlrvGFgq80H5IjolCgUO5oTlwF9kA0eDALd3cVPXHn-G3IQAdNGXO2T63WOQQHR+VBH1Bn9BOCP2gCf8SiEURxHaPggA)

And when it comes to the inference, the situation could be improved by [this PR](https://github.com/microsoft/TypeScript/pull/52972). We can see some improvements already on the TS playground using a TS build from that PR: [TS playground](https://www.typescriptlang.org/play/?ts=5.8.0-pr-52972-10#code/C4TwDgpgBA0hIDUCGAbArtAvFA3lA2gNbwBcUAzsAE4CWAdgOYC6ZSdIUAvgNwBQvEAB5gA9lWBRQkKABUIlAGI1BwNFQgAeAEoAaKAEEqDclCHAIdACYm4iVBgB8UbAApeUKEiPkyh4zvcoNHIIMhcqMi0ASmcnAAUqEQBbGhCNADcRGksHAJjMJzYQPl4paCUVNXkNGVMVC2tcTmcmvTjaswaTHGbsHqc+wPwYKHooYhARADModqYAfhJAjwAfWXlgCtV1DXbhpj1agDJZmQdlqDX8OUVlbc09mAPZKBP23NxJGiSIETRgRZQOhoJIAIwgVEuQSsECm9Agli4TD4zROOCGIzGE2mLyQJgAooIAMboSyaGB6bEzd4LJYeVbrW6VHYyfaHV6nc70qHXDZbKo1NkvN5nPR4YDfX7-QHAsEQqFoGFwugIpEokr0cxUKZIInQG7AGTgTQGvwmTpWGzwZDoCBOdEeC2WGp1cyWpoONzcuHM+Rkfnqcg1Q4bM25QJRMgGo2QGqh7wcs4o-iaiE6vWzJAMaAO0zpexIcxhSgRCjUegMfLxRIpNKKwh0EQAdzoDmTvDJJK80CJIjolCgoLxoUZhuNGjwYCzI7i064bf4AHpF6M6FMISYAAYAOidE84Dk3vCHIV39SsLlzy6gAD15oEp9nWOQQHQiVAXAB9KutXicKIlNe9DrlQW5nm6zr9EeJ4QOBDSXoE153oESQgAGI54q+75fj+PR6NewAABapBQhF-CgiJNmIhB6AAkgA5CgKCrukGw0AwhbQDQwB-gBS4rsBG5QDue5Qcew5wReV4rshHiPhhL5vh+36xL+Hioehz5YcpuGcPhK5ESR5BkWgFFQFRVA0VADFMSxbEceYow8f+gECWuG68NeInnpBnkrtyGl3FUZD1o2LZ8NeB7QRJToIR4SH3nJ05aUpOGqXhUDXr2dDmCoAC0IT9txNCsShaFBeoKXYbp+lQNluXAHl9CFeQxWlS5-GriB5B+cJkm+deAXlb6IV0A2zZ0BFK5ReJp6xdJt6JVA8lVR+NWZSu9VmE1-YWK1Eqlepw33KtLjrVlfYNTtLVtRAvGuV1Hlef1E69UNmnQuN4W9TNMH9XFG2LWVH2Yal52bZd23NXtt33Z1gmgX1om-TFPkAwlwMVQp2lfnowQji4XgMGQsrglQP6ZNk4N1ZD+U3Qdd0uUAA). Based on those findings, I need to take a look at what happens with context-sensitive functions there because they seem to brak this inference.

Note that on the TS playground above I also replaced multiple `extends KeyValue` with `extends {}`. This is a safer default because TS uses the constraint when there are no inference candidates. Through `extends KeyValue` `any`s can easily creep into the user types and I think it's best to avoid that risk.

TLDR: I created a PR with a fix for this in Playwright [here](https://github.com/microsoft/playwright/pull/33782)
